### PR TITLE
Inline CSS with Critters Plugin

### LINKED ISSUE #637: Inline critical css
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behaviour?**
Currently we ship `<link rel=stylesheet ` the webpack's bundle stylesheet, which has to be parsed before first paint/layout.

**If the current behaviour is a bug, please provide the steps to reproduce.**
- serve any preact-cli app

**What is the expected behaviour?**
We can use `critters` to inline just the critical css and then defer load the `bundle.css`

**If this is a feature request, what is motivation or use case for changing the behaviour?**
Web perf

**Please mention other relevant information.**
- Node version: any
- npm version: any
- Operating system: next
- CLI version: any
- Browser: any

### PR DESCRIPTION
**Pull request by lukeed**
**What kind of change does this PR introduce?**

Feature

**Did you add tests for your changes?**

No

**Summary**

Uses [Critters](https://github.com/GoogleChromeLabs/critters) to inline critical CSS on the client bundle(s) for production _and_ development.

The default settings are good 👌 but can be changed by end-user thru `preact.config.js` if/when needed.

Closes #637 

**Does this PR introduce a breaking change?**

No

### CODE REVIEW

**Comment by prateekbh** [comment]
Quick thing, I'd like the bundle.css to defer loaded since this is inlined anyways. WDYT?

**Comment by lukeed** [comment]
Unless I've missed something, both `defer` and `async` are `<script>`-specific attributes... they won't do anything (for now) on other tags.

The default Critters setup uses `rel=preload as=style` which is as close to `defer` as you can get me thinks, before bringing in any JS

**Comment by prateekbh** [comment]
Hmmm, how about we inline what critters give us and put bundle.css as 
`<line rel='defer-stylesheet' href='...'>` and use js to load this later on?

**Comment by lukeed** [comment]
I think our first-paint is already super fast with default setting. Don't think we should be doing anything custom/beyond what Critters offers, as that's the point/goal of Critters, no?

Either way, sounds like you may want the [`{ preload: 'swap' }`](https://github.com/GoogleChromeLabs/critters#preloadstrategy) option.

**Comment by prateekbh** [comment]
Oh I just read: 
>**default:** Move stylesheet links to the end of the document and insert preload meta tags in their place.

This seems fine, if it is indeed moving it to the end. Please just check that we don't have double preload css tags with `--preload`

**Comment by prateekbh** [comment]
Also, tried this right now and running into: https://github.com/GoogleChromeLabs/critters/issues/5

File: packages/cli/lib/lib/webpack/webpack-client-config.js [hunk 1]
```diff
@@ -75,6 +76,7 @@ function clientConfig(env) {
 		},
 
 		plugins: [
+			new CrittersPlugin(),
```
Comment (prateekbh) [inline packages/cli/lib/lib/webpack/webpack-client-config.js]: this should be after `...RenderHTMLPlugin(env),` to resolve the error from https://github.com/GoogleChromeLabs/critters/issues/5

**Comment by prateekbh** [comment]
btw there should be an opt-out/opt-in flag for critters because based on your css it is running into edge cases.

**Comment by lukeed** [comment]
Fixed the plugin ordering 👍 

There _are_ double `link rel=preload` tags, but that shouldn't matter? It still only fires off a single request to `bundle.*.css` & I'm not getting any unused warnings.

screen shot 2018-09-11 at 5 29 40 pm

**Comment by lukeed** [comment]
I think I see what you're talking about. 

The `default` mode append a `<link rel=stylesheet href="bundle.*.css"/>` to the end of the document – is that what you were referring to?

The tag is removed with `js` and `js-lazy`, but has no effect on the Lighthouse score w/ our Preact templates since the CSS is so little.

**Comment by prateekbh** [comment]
>The tag is removed with js and js-lazy, but has no effect on the Lighthouse score w/ our Preact templates since the CSS is so little.

This has affect when the app grows e.g. material.preact.com

preaload: swap works pretty perfect with just a few characters even for bigger sites.

>There are double link rel=preload tags, but that shouldn't matter?

Chrome throws a warning for this in console IIRC

**Comment by lukeed** [comment]
Gotcha. It warns if unused, but not on double preload tag.

Feel free to change it to swap, I don't have any objections! I'll likely be away from desk for a few hours

**Comment by prateekbh** [comment]
How about along with a opt-out flag?

**Comment by lukeed** [comment]
Personally I don't like opt-out flags, so I'd be hesitant. I also not sure why you'd need one?

Arguably, if we're making something a default behavior I feel that we should stand behind it 100©

Unlike the esm teacher, there's no compatibility or support issues.

**Comment by prateekbh** [comment]
Because I feel that it's not very stable, I made like 4-5 changes to the source code in my node_modules before I could get it to work

**Comment by prateekbh** [comment]
@lukeed @developit should this be concerning that even after using preload(swap) lighthouse still complains about unused for the link tag and even the inlined css is being reported as extra css supplied

e.g. https://www.webpagetest.org/lighthouse.php?test=180913_MG_dc5a8260f5c19115230143e77c18527d&run=1

**Comment by prateekbh** [comment]
Also, another use case of opt-out flag/question: if I use a case in July aren't they better at giving me my critical css rather than critters?

**Comment by lukeed** [comment]
If it's not stable, then:

a) we shouldn't include it
b) we should fix it so that it's stable; or
c) _enable_ the feature with a flag, not disable with flag

**Comment by lukeed** [comment]
Oops, sorry, I meant that those were three possible choices; not to do all three 😅 

I'd be totally stoked with inline-css as a default feature, but we need to be confident in it. If we're not, then it either can't be a default behavior or shouldn't be there at all

**Comment by prateekbh** [comment]
+1, does an opt in flag sound like a decent approach now?

**Comment by ForsakenHarmony** [comment]
We coulc do it like the esm stuff

Opt in now and if we solve all the problems we can make it the default

**Comment by prateekbh** [comment]
added a flag, and conditional preload of css if critters is used

File: packages/cli/lib/index.js [hunk 2]
```diff
@@ -31,7 +31,8 @@ prog
 	.option('--analyze', 'Launch interactive Analyzer to inspect production bundle(s)')
 	.option('--prerenderUrls', 'Path to pre-rendered routes config', 'prerender-urls.json')
 	.option('-c, --config', 'Path to custom CLI config', 'preact.config.js')
-	.option('--esm', '[EXPERIMENTAL] Builds ES-2015 bundles for your code.', false)
+	.option('--esm', 'Builds ES-2015 bundles for your code.', false)
+	.option('--critical-css', 'Adds critical cssto the prerendered markup.', true)
```
Comment (lukeed) [inline packages/cli/lib/index.js]: I would make this `--inline` or `--inline-css`, but maybe that's just me /shrug

File: packages/cli/lib/resources/template.html [hunk 3]
```diff
@@ -10,8 +10,9 @@
 		<% if (htmlWebpackPlugin.options.manifest.theme_color) { %>
 			<meta name="theme-color" content="<%= htmlWebpackPlugin.options.manifest.theme_color %>">
 		<% } %>
+		<% const filesRegexp = options.criticalCss ? /\.(js)$/ : /\.(css|js)$/;%>
```
Comment (lukeed) [inline packages/cli/lib/resources/template.html]: I don't understand – we'd still want to preload _some_ CSS, especially the routes' chunks.

File: packages/cli/lib/index.js (hunk 2 above)
Comment (prateekbh) [reply packages/cli/lib/index.js]: Nah that sounds better

**Comment by lukeed** [comment]
@prateekbh FYI - failing tests are now actually relevant to your changes

**Comment by prateekbh** [comment]
fixed everything, good to merge. plz review

**Comment by developit** [comment]
Critters issue should be fixed in [1.2.0](https://github.com/GoogleChromeLabs/critters/releases/tag/1.2.0).

**Comment by prateekbh** [comment]
@developit I have no idea why critters 1.2.0 complains JSDOM is not a constructor, hence keeping this PR @ 1.1.0 and will upgrade when we fix that

**Comment by prateekbh** [comment]
@ForsakenHarmony can you review this one :)

File: packages/cli/lib/resources/template.html [hunk 4]
```diff
@@ -10,8 +10,9 @@
 		<% if (htmlWebpackPlugin.options.manifest.theme_color) { %>
 			<meta name="theme-color" content="<%= htmlWebpackPlugin.options.manifest.theme_color %>">
 		<% } %>
+		<% const filesRegexp = htmlWebpackPlugin.options.inlineCss ? /\.(chunk\.\w{5}\.css|js)$/ : /\.(css|js)$/;%>
```
Comment (reznord) [inline packages/cli/lib/resources/template.html]: `template.html` is already complicated and messed up. @developit any plans on making it similar to what CRA does?

Comment (prateekbh) [reply packages/cli/lib/resources/template.html]: Well I can take a look at this in next iteration, but not for this PR

**Pull request merged**